### PR TITLE
ceph-volume/batch: fail on filtered devices when non-interactive

### PR DESCRIPTION
**Pull request by jan--f**
When batch is called non-interactively and a user explicitly specifies,
say a db-device, this will be filtered when unavailable. This can cause
the resulting OSD to be very different from the users intention
(standalone vs external db when the db-device was filtered). If devices
get filtered in non-interactive mode, ceph-volume should fail.

Fixes: https://tracker.ceph.com/issues/43105

Signed-off-by: Jan Fajerski <jfajerski@suse.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by jan--f** [comment]
jenkins test ceph-volume tox

**Comment by jan--f** [comment]
jenkins test ceph-volume all

**Comment by jan--f** [comment]
One tox failure is addressed in #31979.

@alfredodeza The tox test also has a unicode decode failure again, mind taking a look?

**Comment by jan--f** [comment]
ping @sebastian-philipp

**Comment by sebastian-philipp** [comment]
Hm. We're already filtering devices in cephadm now in #32545 . I'd be :+1: to just bail out immediately, if _any_ device is not available. Any objections @leseb ?

**Comment by jan--f** [comment]
> Hm. We're already filtering devices in cephadm now in #32545 . I'd be +1 to just bail out immediately, if _any_ device is not available. Any objections @leseb ?

Hmm the intention is to get batch idempotent, so I'd rather not just fail when any unavailable drive is passed.

**Comment by sebastian-philipp** [comment]
At least when using drive Groups, we're already loosing idempotency. Maybe we'rebgoing seeing some unexpected interactions between python-common's drive groups and c-v's idempot algorithm

**Comment by jan--f** [comment]
> At least when using drive Groups, we're already loosing idempotency.

Not sure I understand why? If DG are already filtering so that only available drives make it to c-v then there is no harm done? Idempotency is only achievable is the call is fully specified (i.e. sizes are passed and such) and at that point it makes no difference if an unavailable drive is filtered out before it gets to c-v.
What we do gain however is less surprises for users running ceph-volume directly..

**Comment by jschmid1** [comment]
I agree with @jan--f 's proposal here. It's about the non-interactive mode only, which should give the user immediate feedback instead of magically hiding disks. For the fully automated, drivegroup based call we're basically always calling the same command which should suffice for achieving idempotency.

**Comment by sebastian-philipp** [comment]
🤷

File: src/ceph-volume/ceph_volume/tests/devices/lvm/test_batch.py [hunk 1]
```diff
@@ -65,7 +65,7 @@ def test_disjoint_device_lists(self, factory):
         b._filter_devices()
         with pytest.raises(Exception) as disjoint_ex:
             b._ensure_disjoint_device_lists()
-        assert 'Device lists are not disjoint' in str(disjoint_ex.value)
+            assert 'Device lists are not disjoint' in str(disjoint_ex.value)
```
Comment (sebastian-philipp) [inline src/ceph-volume/ceph_volume/tests/devices/lvm/test_batch.py]: this change looks wrong to me

Comment (jan--f) [reply src/ceph-volume/ceph_volume/tests/devices/lvm/test_batch.py]: care to elaborate?

`_ensure_disjoint_device_lists()` raises and Exception when called with overlapping device lists. The assert checks for the certain message in the assertion.

Comment (sebastian-philipp) [reply src/ceph-volume/ceph_volume/tests/devices/lvm/test_batch.py]: https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions

Comment (jan--f) [reply src/ceph-volume/ceph_volume/tests/devices/lvm/test_batch.py]: > https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions

Not actually an explanation. So turns out with, being just syntactic sugar doesn't open a new scope. In other words either indentation is fine.

Comment (sebastian-philipp) [reply src/ceph-volume/ceph_volume/tests/devices/lvm/test_batch.py]: actually no:

```python
(venv) ➜  foo python
Python 3.6.9 (default, Nov  7 2019, 10:44:02) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pytest
>>> with pytest.raises(Exception) as ex:
...     raise Exception("foo")
...     assert 'bar' in str(ex.value)
# passes
```

problem is, as soon as we leave the indented block, the assert is no longer executed.  The scope of variables in Python is as bad as you can imagine, the control flow of statements is not as bad.

Comment (jan--f) [reply src/ceph-volume/ceph_volume/tests/devices/lvm/test_batch.py]: oh right...syntactic sugar was the hint :)

**Comment by jan--f** [comment]
retest this please

**Comment by jan--f** [comment]
jenkins test ceph-volume tox

**Pull request merged**